Index: linux/arch/ppc/kernel/asm-offsets.c
===================================================================
--- linux.orig/arch/ppc/kernel/asm-offsets.c	2006-09-28 15:02:03.000000000 +0200
+++ linux/arch/ppc/kernel/asm-offsets.c	2006-09-28 15:02:17.000000000 +0200
@@ -26,6 +26,7 @@
 #include <asm/cputable.h>
 #include <asm/thread_info.h>
 #include <asm/vdso_datapage.h>
+#include <hpc/task.h>
 
 #define DEFINE(sym, val) \
 	asm volatile("\n->" #sym " %0 " #val : : "i" (val))
@@ -171,5 +172,12 @@
 	DEFINE(NSEC_PER_SEC, NSEC_PER_SEC);
 	DEFINE(CLOCK_REALTIME_RES, TICK_NSEC);
 
+#ifdef CONFIG_OPENMOSIX
+	DEFINE(TASK_om, offsetof(struct task_struct, om));
+	DEFINE(OM_dflags, offsetof(struct openmosix_task, dflags));
+	BLANK();
+	DEFINE(DDEPUTY, DDEPUTY);
+	DEFINE(DREMOTE, DREMOTE);
+#endif /* CONFIG_OPENMOSIX */
 	return 0;
 }
