diff --exclude=.git -Nru linux-2.6.28.7/arch/x86/include/asm/tmig32.h linux-2.6.28.7-lpmi/arch/x86/include/asm/tmig32.h
--- linux-2.6.28.7/arch/x86/include/asm/tmig32.h	1970-01-01 01:00:00.000000000 +0100
+++ linux-2.6.28.7-pms/arch/x86/include/asm/tmig32.h	2009-03-06 19:59:09.000000000 +0000
@@ -0,0 +1,45 @@
+/*
+ *	Copyright (C) 2002-2004 Moshe Bar <moshe@moshebar.com>
+ *	Copyright (C) 2005-2006 Vincent Hanquez <vincent@snarc.org>
+ *     Copyright (C) 2007-2009 Julia Longtin <juri@solarnetone.org>
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published
+ * by the Free Software Foundation; version 2 only.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ * 
+ * Original Mosix code Copyright (C) Amnon Barak, Amnon Shiloh
+ *
+ * Changes for <=2.6.17 by Vincent Hanquez and Alexander Nyberg
+ * Changes for 2.6.28.7 by g_remlin <g_remlin@rocketmail.com>
+ * Additional changes by Julia Longtin <juri@solarnetone.org>
+ *
+ */
+
+#ifdef CONFIG_X86
+#ifndef _TMIG_ARCH_TMIG32_H
+#define _TMIG_ARCH_TMIG32_H
+
+#define NR_MAX_SYSCALL_ARG	6
+
+static inline long arch_get_sys_arg(unsigned int n, struct pt_regs *regs)
+{
+	BUG_ON(n >= NR_MAX_SYSCALL_ARG);
+	return *(((long *) regs) + n);
+}
+
+static inline unsigned long arch_get_sys_nb(struct pt_regs *regs)
+{
+	return regs->ax;
+}
+
+#define ARCH_TASK_GET_USER_REGS(p)					\
+	(((struct pt_regs *)						\
+	(THREAD_SIZE - 8 +(unsigned long) current_thread_info())) - 1)
+
+#endif /* _TMIG_ARCH_TMIG32_H */
+#endif /* CONFIG_X86 */
